Apps/Src/ActivateMotor.c

```diff
@@ -0,0 +1,118 @@
+
+#include "ActivateMotor.h"
+
+
+static void motor_startup(OS_ERR *err) {
+    Precharge_Write(MOTOR_PRECHARGE, ON); // Activate the motor recharge
+
+    OSTimeDlyHMSM(0, 0, 5, 0, OS_OPT_TIME_HMSM_STRICT, &err);
+    // TODO: check for errors
+
+    Contactors_Set(MOTOR, ON); // Actually activate motor contactors
+    Precharge_Write(MOTOR_PRECHARGE, OFF);
+}
+
+
+void Task_MotorConnection(void *p_arg) {
+    car_state_t *car_state = (car_state_t *) p_arg;
+
+    OS_ERR err;
+    CPU_TS ts;
+
+    Contactors_Init(MOTOR);
+    motor_startup(&err);
+
+    // Now we need to spawn the four CAN related tasks
+
+    // Spawn SendCarCAN
+    OSTaskCreate(
+        (OS_TCB*)&SendCarCAN_TCB,
+        (CPU_CHAR*)"SendCarCAN",
+        (OS_TASK_PTR)Task_SendCarCAN,
+        (void*)NULL,
+        (OS_PRIO)TASK_SEND_CAR_CAN_PRIO,
+        (CPU_STK*)SendCarCAN_Stk,
+        (CPU_STK_SIZE)WATERMARK_STACK_LIMIT,
+        (CPU_STK_SIZE)TASK_SEND_CAR_CAN_STACK_SIZE,
+        (OS_MSG_QTY)NULL,
+        (OS_TICK)NULL,
+        (void*)NULL,
+        (OS_OPT)(OS_OPT_TASK_STK_CLR),
+        (OS_ERR*)&err
+    );
+    // TODO: check for task creation error
+
+    // Spawn ReadCarCAN
+    OSTaskCreate(
+        (OS_TCB*)&ReadCarCAN_TCB,
+        (CPU_CHAR*)"ReadCarCAN",
+        (OS_TASK_PTR)Task_ReadCarCAN,
+        (void*)NULL,
+        (OS_PRIO)TASK_READ_CAR_CAN_PRIO,
+        (CPU_STK*)ReadCarCAN_Stk,
+        (CPU_STK_SIZE)WATERMARK_STACK_LIMIT,
+        (CPU_STK_SIZE)TASK_READ_CAR_CAN_STACK_SIZE,
+        (OS_MSG_QTY)NULL,
+        (OS_TICK)NULL,
+        (void*)NULL,
+        (OS_OPT)(OS_OPT_TASK_STK_CLR),
+        (OS_ERR*)&err
+    );
+    // TODO: check for task creation error
+
+    // Spawn ReadTritium
+    OSTaskCreate(
+        (OS_TCB*)&ReadTritium_TCB,
+        (CPU_CHAR*)"ReadTritium",
+        (OS_TASK_PTR)Task_ReadTritium,
+        (void*)NULL,
+        (OS_PRIO)TASK_READ_TRITIUM_PRIO,
+        (CPU_STK*)ReadTritium_Stk,
+        (CPU_STK_SIZE)WATERMARK_STACK_LIMIT,
+        (CPU_STK_SIZE)TASK_READ_TRITIUM_STACK_SIZE,
+        (OS_MSG_QTY)NULL,
+        (OS_TICK)NULL,
+        (void*)NULL,
+        (OS_OPT)(OS_OPT_TASK_STK_CLR),
+        (OS_ERR*)&err
+    );
+
+    // TODO: check for task creation error
+
+    // Spawn SendTritium
+    OSTaskCreate(
+        (OS_TCB*)&SendTritium_TCB,
+        (CPU_CHAR*)"SendTritium",
+        (OS_TASK_PTR)Task_SendTritium,
+        (void*)NULL,
+        (OS_PRIO)TASK_SEND_TRITIUM_PRIO,
+        (CPU_STK*)SendTritium_Stk,
+        (CPU_STK_SIZE)WATERMARK_STACK_LIMIT,
+        (CPU_STK_SIZE)TASK_SEND_TRITIUM_STACK_SIZE,
+        (OS_MSG_QTY)NULL,
+        (OS_TICK)NULL,
+        (void*)NULL,
+        (OS_OPT)(OS_OPT_TASK_STK_CLR),
+        (OS_ERR*)&err
+    );
+    // TODO: check for task creation error
+
+    while (1) {
+        // Wait until some change needs to be made to the motor state
+        OSSemPend(&MotorConnectionChange_Sem4, 0, OS_OPT_PEND_BLOCKING, &ts, &err);
```

Comment:
Why is this pending a semaphore? I thought this task was spawned by the Switches task. It should just kill itself after completing its purpose and wait to be spawned again. 

Essentially, this should act as a regular function. The reason I would recommend keeping it as a task, though, is because of the 5 second delay. If this were a regular function call from the `ReadSwitches` task, then that task would be blocked for 5 seconds while it waited for this to finish.

Reply:
We wanted to have the ability to turn the motor on or off, depending on the state of the key. Read Switches checks where the key is, and if it decides that the state of the motor should change, then it posts to the semaphore. If we just killed the task when it was done, there would be no way to turn the motor on again more than once.

Reply:
Well I thought that's what the state variable was for. To communicate to this task whether it should turn it on or off. That doesn't require a semaphore or a blocked task as far as I can see

Reply:
But if the task continuously checked the state variables for changes, it would just chew up cpu time. The semaphore lets it know that something has changed, at which point it can check the state variables and figure out what do. Another option would also be to respawn the task every time we want to turn the motor on, and have a function to turn it off.

Reply:
I'm not saying to continuously check. What I mean is that the Switches task turns this task on when something changes (whether that's off->on or on->off). This task then runs once (no looping). It first checks the state variable to know whether it's turning the motor on or off. Depending on which one, it either spawns the new tasks or kills those tasks and then either turns the motor on or turns it off. No need for a loop or a blocked task.

Reply:
I think I agree with @Cam0Cow here. A blocked task doesn't take up any resources that we need anyways, and it's going to be quicker to unblock and manipulate the motor state than to spawn a new task every time.

Reply:
Okay well then we shouldn't have it do both. Either the switches task just blocks this one or it spawns it. In its current state it spawns it once and then blocks it.

Reply:
Maybe interesting to look into using a task semaphore instead (as opposed to a regular semaphore). See pg. 267 here https://www.analog.com/media/en/dsp-documentation/software-manuals/Micrium-uCOS-III-UsersManual.pdf

Reply:
That looks like exactly what we need @chaseblock